Fix #74: Add Checked.int8/uint8 and Nullable.int8/uint8/single/double

dsyme opened the pull request:
[Tracking UserVoice entry](http://fslang.uservoice.com/forums/245727-f-language/suggestions/6995974-add-checked-int8-uint8-and-nullable-int8-uint8-sin)

F# defines "int8" and "uint8" as synonyms for "sbyte" and "byte", both as types and operators

While reviewing the behaviour of "open Checked" (which brings new versions of operators into scope) I noticed that new versions of the "int8" and "uint8" operators are not defined in the "Checked" module.  This is inconsistent, though a simple workaround is to use the checked "sbyte" and "byte" operators instead.

I also noticed that "Nullable.int8" and "Nullable.uint8" are not defined.

All this should be fixed and the fix is simple.  I don't mind if it is in FSharp.Core 4.4.0.0 or later.

Cheers
Don

mexx commented on src/fsharp/FSharp.Core/fslib-extra-pervasives.fsi:
```diff
@@ -92,15 +92,31 @@ module ExtraTopLevelOperators =
-    [<CompiledName("ToByte")>]
-    val inline uint8       : value:^T -> byte       when ^T : (static member op_Explicit : ^T -> byte)       and default ^T : int        
+    [<CompiledName("ToUInt8")>]
+    val inline uint8       : value:^T -> uint8       when ^T : (static member op_Explicit : ^T -> byte)       and default ^T : int        
```
Should `static member op_Explicit : ^T -> byte` not be also `static member op_Explicit : ^T -> uint8` for consistency?

mexx commented on src/fsharp/FSharp.Core/fslib-extra-pervasives.fsi:
```diff
@@ -92,15 +92,31 @@ module ExtraTopLevelOperators =
     /// primitive numeric types. For strings, the input is converted using <c>Byte.Parse()</c> on strings and otherwise requires a <c>ToByte</c> method on the input type.</remarks>
-    [<CompiledName("ToByte")>]
-    val inline uint8       : value:^T -> byte       when ^T : (static member op_Explicit : ^T -> byte)       and default ^T : int        
+    [<CompiledName("ToUInt8")>]
```
Do I get it right, this will be a breaking change for external usages?

dsyme replied:
I have reverted this change.  The compiled name should really always have been "ToUInt8", but you are correct, we won't change it now for backwards-compat reasons.

dsyme commented:
I've updated the PR with 
* more testing (including for the added nullable ops)
* Nullable.double and Nullable.single, which wewre also missing.

I'll write up a mini-spec for this too.

latkin commented on src/fsharp/FSharp.Core/Linq.fsi:
```diff
@@ -109,6 +117,15 @@ namespace Microsoft.FSharp.Linq
+        /// static conversion method on the input type.</summary>
+        /// <param name="value">The input value.</param>
+        /// <returns>The converted sbyte</returns>
+        [<CompiledName("ToInt8")>]
```
If new APIs are being added, you will need to update the surface area tests in FSharp.Core.Unittests.  Note that ATM this test is failing, as some required updates were missed during recent PRs. @KevinRansom I think is working on fixing it.

latkin commented:
This looks like a combo mini-feature and cleanup/refactor. Do I understand correctly?

If so, I will direct you to our contribution guidelines, written by a wise contributor to this repo :smile: 
> Code cleanup should generally be made separate to other checkins where possible, or the parts that are cleanup should be labelled as such.

dsyme commented on src/fsharp/FSharp.Core/Linq.fs:
```diff
@@ -302,30 +319,41 @@ module LeafExpressionConverter =
-    let (|ConvUInt16Q|_|)      = (|SpecificCallToMethod|_|) (methodhandleof (fun x -> uint16 x))
-    let (|ConvUInt32Q|_|)      = (|SpecificCallToMethod|_|) (methodhandleof (fun x -> uint32 x))
-    let (|ConvUInt64Q|_|)      = (|SpecificCallToMethod|_|) (methodhandleof (fun x -> uint64 x))
+    let (|ConvCharQ|_|)        = (|SpecificCallToMethod|_|) (methodhandleof (fun x -> Operators.char x))
```
Throughotu this file "Operators." has been included to give important code clarification as part of the bug fix as to which operator implementation is being called

dsyme commented:
great, thanks

- Yes, this is a feature.

- I've reduced the code change to what's needed for a clean, clarifying bugfix

- I've added a UserVoice entry (since all features need one)

- I've updated the surface area tests

enricosada commented on src/fsharp/FSharp.Core/Linq.fs:
```diff
@@ -77,53 +77,66 @@ module NullableOperators =
 
         [<CompiledName("ToInt32")>]
-        let inline int32 (x:Nullable<_>) = if x.HasValue then Nullable(int32 x.Value) else Nullable()
+        let inline int32 (x:Nullable<_>) = if x.HasValue then Nullable(Operators.int32 x.Value) else Nullable()
```
why `Nullable.int16` parameter is `x: Nullable< ^T >` but `Nullable.int32` is `x: Nullable<_>` ?

dsyme replied:
It doesn't matter either way. However I won't normalize it as part of this PR since it's unrelated cleanup.

enricosada replied:
@dsyme, sure, i was curious to learn if there was a diff behaviour in this context or only code style. ty

latkin commented:
Hmm, since this will alter the public surface area of FSharp.Core, it really needs to be done for 4.0, or else wait for the next version bump (might be a while). Seems innocuous enough... @KevinRansom what do you think, should we pull this for 4.0?

KevinRansom commented:
Absolutely, good eyes on spotting the inconsistency.
